flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBaseTest.java

```diff
@@ -218,23 +221,127 @@ public void go() throws Exception {
 			Thread.sleep(5);
 		}
 
-		Assert.assertEquals(1, casSinkFunc.getNumOfPendingRecords());
+		Assert.assertEquals(1, casSinkFunc.getAcquiredPermits());
 		completableFuture.complete(null);
-		Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+		Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+	}
+
+	@Test(timeout = 5000)
+	public void testAcquireOnInvoke() throws Exception {
+		final CassandraSinkBaseConfig config = CassandraSinkBaseConfig.newBuilder()
+			.setMaxConcurrentRequests(1)
+			.setMaxConcurrentRequestsTimeout(5, TimeUnit.MILLISECONDS)
+			.build();
+		try (TestCassandraSink casSinkFunc = new TestCassandraSink(config)) {
+			casSinkFunc.open(new Configuration());
+
+			CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+
+			casSinkFunc.enqueueResultSetFuture(ResultSetFutures.fromCompletableFuture(completableFuture));
+
+			Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+
+			casSinkFunc.invoke("hello");
+
+			Assert.assertEquals(0, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(1, casSinkFunc.getAcquiredPermits());
+
+			completableFuture.complete(null);
+		}
+	}
+
+	@Test(timeout = 5000)
+	public void testReleaseOnSuccess() throws Exception {
+		final CassandraSinkBaseConfig config = CassandraSinkBaseConfig.newBuilder()
+			.setMaxConcurrentRequests(1)
+			.setMaxConcurrentRequestsTimeout(5, TimeUnit.MILLISECONDS)
+			.build();
+		try (TestCassandraSink casSinkFunc = new TestCassandraSink(config)) {
+			casSinkFunc.open(new Configuration());
+
+			CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+
+			casSinkFunc.enqueueResultSetFuture(ResultSetFutures.fromCompletableFuture(completableFuture));
+
+			Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+
+			casSinkFunc.invoke("hello");
+
+			Assert.assertEquals(0, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(1, casSinkFunc.getAcquiredPermits());
+
+			completableFuture.complete(null);
+
+			Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+			Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+		}
+	}
+
+	@Test(timeout = 5000)
+	public void testReleaseOnFailure() throws Exception {
+		final CassandraSinkBaseConfig config = CassandraSinkBaseConfig.newBuilder()
+			.setMaxConcurrentRequests(1)
+			.setMaxConcurrentRequestsTimeout(5, TimeUnit.MILLISECONDS)
+			.build();
+		TestCassandraSink casSinkFunc = new TestCassandraSink(config);
+
+		casSinkFunc.open(new Configuration());
+
+		CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+
+		casSinkFunc.enqueueResultSetFuture(ResultSetFutures.fromCompletableFuture(completableFuture));
+
+		Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
```

Comment:
the majority of these tests could be moved into a shared method to reduce code duplication.